### PR TITLE
drivers: flashdisk: implement basic write caching

### LINKED ISSUE #30212: Disk rewrites same flash page multiple times.
While testing USB mass storage on nrf52 I noticed that the write performance was very bad. The reason for this is because disk flash uses a sector size of 512 byte but the erase size of flash is 4096 bytes. For each written sector the disk_access_flash driver reads 4096 bytes, updates 512 bytes, erases the flash pages and writes it. This is highly inefficient since writing multiple contiguous sectors is a common operation.

Changing sector size in disk to match the erase flash size is not an option since 512 byte it often needed for compatibility.

I propose to add a cache in disk_access_flash that stores last writes and flushes them when writes to another flash page is issued. USB mass storage class should use the DISK_IOCTL_CTRL_SYNC ioctl call to flush the cache.
I have retrofitted this as a PoC and it seems to work fine, owever it would be good if the whole driver was written with this intention.

I would also recommend to rewrite disk_access_flash so that it instead uses flash_map.h so that the disks can be described in devicetree instead of hardcoded configuration parameters. Doing that would take one step towards unifying the file system backend APIs.

### PR DESCRIPTION
**Pull request by desowin**
Cache written data to avoid rewriting same flash page multiple times when writing subsequent flash pages. The cache is used for reads to account for reading not yet committed (i.e. dirty) page data. Speeding up reads is not intention of this patch and therefore the read path does not modify cache state.

Fixes: #30212

This simple caching quite significantly improves write performance. On nRF52840 DK the multiblock writes (SCSI Write(10) Request) are around 9.5 times faster. The speedup is more than would be expected due to only performing 1/8 of the erase/program cycles as before. The additional performance improvement is likely from reduced time lost to NAKed USB OUT transactions.

Thanks to caching it is now possible to write files greater than 101.5 KiB to nRF52840 DK flashdisk. Without caching the write would take longer than 30 seconds and the Linux host issued USB device reset and restarted the transfer. The restart did not help and eventually host gave up. While caching does not solve the issue completely, it significantly improves the chance of restart to succeed because flash contents are not erased when host writes unmodified data. The 30 seconds SCSI Write(10) Request timeout issue is out of scope of this PR.

### CODE REVIEW

**Comment by Laczen** [comment]
@desowin, a general code remark regarding the thread-safety. I prefer something like:
```
    k_mutex_lock(&ctx->lock, K_FOREVER);
    ...
    rc = whatever...
    if (rc != 0) { /* an error condition */
        goto end;
    }
    ...
    everything ok (then normally rc = 0)
end:
    k_mutex_unlock(&ctx->lock);
    return rc;
```

File: drivers/disk/flashdisk.c [hunk 1]
```diff
@@ -297,31 +329,41 @@ static int disk_flash_access_write(struct disk_info *disk, const uint8_t *buff,
...
 {
+	int rc;
 	struct flashdisk_data *ctx;
 
 	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
 
 	switch (cmd) {
 	case DISK_IOCTL_CTRL_SYNC:
-		return 0;
+		k_mutex_lock(&ctx->lock, K_FOREVER);
```
Comment (Laczen) [inline drivers/disk/flashdisk.c:353]: Move mutex lock/unlock into flashdisk_cache_commit.

Comment (desowin) [reply drivers/disk/flashdisk.c:353]: I don't agree. Zephyr mutex is recursive so acquiring already acquired mutex would not cause issues, but the cache functions are really supposed to be called with the mutex held. Note that `flashdisk_cache_commit()` can also be called from `disk_flash_access_write()` with the lock being held.

Comment (Laczen) [reply drivers/disk/flashdisk.c:353]: Ok, it is also possible to create a flashdisk_ctrl_sync routine that takes the locks, calls cache_commit and unlocks. But you can leave it as is.

**Review by danieldegrasse — APPROVED** [review]

**Review by Laczen — APPROVED** [review]

**Pull request merged**